### ISSUE TITLE
Provide a way to ease building cross-platform single binary distribution

### ISSUE THREAD

**Issue by asticode** [asticode/go-astilectron#13]
Related comments:

- https://github.com/asticode/go-astilectron/issues/11#issuecomment-303970242
- https://github.com/asticode/go-astilectron/issues/10#issuecomment-303819875

**Comment by veecue** [comment]
How is your status on this since it would fix #16 aswell. Any ideas yet?

**Comment by asticode** [comment]
The way I see things, that wouldn't fix #16 as well since #16 is about creating a window without using Electron whereas this issue is about helping those who want to build cross-platform single binary distribution.

Regarding this issue here's where I stand:

- when people build  cross-platform single binaries they usually use another language than GO (bash for instance) to get everything working properly
- the only feature this lib could bring to the table would be to provide a command that would download all dependencies in a specific folder. That way people would be able to `go generate` depending on the target OS/Arch without the need to fetch files on their own

Any thoughts?

**Comment by veecue** [comment]
I think it wouldn't be hard to write a batch script for the job using `wget`. However I don't really have an idea about how to get this done in windows batch.

**Comment by asticode** [comment]
and what would this program do precisely? My main fear is that it would be tough to add custom steps to this program...

**Comment by veecue** [comment]
I have worked on the issue and came up with this rather crude proof of concept: https://github.com/veecue/go-astilectron-bindata

**Comment by asticode** [comment]
This is a very nice proof of concept!

Here are a few comments:
- I think astilectron's`initElectronDownloadSrc` should be exported in a way that it would be usable in this new package instead of rewriting it as `getElectronPath`. Adding a `ElectronPath` function in astilectron might do the trick: it would be used by both `initElectronDownloadSrc` and this new package.
- maybe that would be more readable if templates were used to generate the output file
- instead of using `a` and `e`, I would use self-explanatory constants to avoid any misunderstandings
- instead of creating your own writer, what do you think about using the writer of go-bindata : https://github.com/jteeuwen/go-bindata/blob/master/bytewriter.go#L18

With those changes made, I would love to see a full working example!

Thanks again for working on that!
Cheers

**Comment by veecue** [comment]
Thanks for your feedback!
- I sent a PR (#22)
- I want to stick to streams, because the downloads would have to be cached in RAM when using templates
- Done (`astilectron.zip` and `electron.zip`)
- I tried it, but it creates a slice literal instead of a string, which led to massive RAM leakage of the compiler in my tests

Cheers

**Comment by veecue** [comment]
Well, I'm a student, I got time ;)
 - I've fixed them
 - Done
If you look at my repo at this point, I think that merging it into astilectron would make sense (maybe after the app is improved to support some params, etc.)
Greetings

**Comment by asticode** [comment]
I've merged your PR.

Merging the app in astilectron could make sense, I'll have to think about how to make it smoothly and if it doesn't add unnecessary complexity and should be ketp in a separate repo.

Cheers

**Comment by aashayshah** [comment]
Any updates on this?

**Comment by veecue** [comment]
Everything is working for me so far. Steps:
- `go get github.com/veecue/go-astilectron-bindata`
- add this line to your code: `//go:generate go-astilectron-bindata` and run `go generate` *OR* simply run `go-astilectron-bindata`
- add `"github.com/veecue/go-astilectron-bindata"` to your includes
- add this code before `a.Start()`: `a.SetProvisioner(astilectron_bindata.NewProvisioner(Disembed))`

**Comment by asticode** [comment]
For now I think this is indeed the best solution. I don't want to merge this logic in `astilectron` until it has a wider purpose such as bundling the binary as well (for darwin systems for instance) and until I'm sure it doesn't add unnecessary complexity to `astilectron`. Which could take time.

Cheers

**Comment by asticode** [comment]
@veecue I've thought about this issue and I've changed my mind, I think now is the time to merge your repo into astilectron if you're still up to it.

Here's what I've in mind:

- create a `cmd` folder at the root level, then a `astilectron-bindata` folder into it and then the content of your `go-astilectron-bindata/app.go` into a `main.go` file
- your `NewProvisioner` should go into the `provisioner.go` file as `func NewDefaultDisembedderProvisioner(d Disembedder) Provisioner`
- update the `5.single_binary_distribution` example to show how to use this new feature (you'll need to update the example section in the readme as well to let people known that they'll need to run the cmd before running the exampl)
- add a `Commands` section in the readme before the `I want to see it in actions!` section, and add a sub-section `Bindata` that explain how to use this new feature

What do you think?

**Comment by veecue** [comment]
I'll start working on it. However, I noticed one problem: I use the `github.com/ogier/pflag` package because the `flag` package doesn't support POSIX-style flags with shortcuts (what where the go devs thinking???). Do you mind two different flag packages used in one repo?

**Comment by asticode** [comment]
I'd rather limit the imports and have only one flag package (the `flag` one) in the repo if that's okay with you

**Comment by veecue** [comment]
It's okay. Should I use the short forms `c` oder the long forms `current-only` then?

**Comment by asticode** [comment]
I prefer the short form `c`

**Comment by veecue** [comment]
Okay, fixed that one. How would you like to handle licensing and merging the repos?

**Comment by asticode** [comment]
I was thinking you could make a PR with the changes I mentioned in a previous comment (https://github.com/asticode/go-astilectron/issues/13#issuecomment-317169253). Did you have something else in mind?

**Comment by rand99** [comment]
as a bundle helper ...

for osx there is this tool to generate an app with icon (go) 
https://github.com/murlokswarm/macpack
maybe this code could be used

**Comment by asticode** [comment]
@veecue did you have time to think about this?

**Comment by asticode** [comment]
@veecue I had a change of heart regarding this issue. Let me explain:

Following #38, I've created [go-astilectron-bundler](https://github.com/asticode/go-astilectron-bundler) which provides a way to bundle an Astilectron app properly (starting with the proper MacOS bundle structure).

I think that's what I had in mind when I created this issue:
- **go-astilectron** that would provide a way to communicate with **astilectron**
- **go-astilectron-bootstrap** that would provide an easier and more rigurous way to use **go-astilectron**
- **go-astilectron-bundler** that would provide an easy way to bundle an Astilectron app for any specified os/arch couple.

I think your project would make more sense in [go-astilectron-bundler](https://github.com/asticode/go-astilectron-bundler) and more specifically https://github.com/asticode/go-astilectron-bundler/blob/master/bundler.go#L155.

This new project is still a work in progress but I've the feeling it's the way to go.

What do you think?

**Comment by veecue** [comment]
Hey @asticode,
this sounds really promising. Do you aim to create macOS apps, windows installers and linux packages for all distros? I really appreciate this project. I am still hoping for golang to provide a proper way to embed binary resorces, but until then, we could integrate my project. I could also test your new repo in a great way, by trying to compile GroupMatcher :)
TL;DR: I'm interested in the developement
Cheers, veecue

**Comment by asticode** [comment]
For now I'm aiming at something simpler than installers:
- for MacOS, create a bundle structure (creating the proper folders and files)
- for Linux and Mac, add an icon to the exec

I've also decided to simplify the bootstrap and move it in a separate repo (https://github.com/asticode/go-astilectron-bootstrap) so that we can give a more structured way of using **Astilectron** to its users.

Do you mind closing the PR you've done in this repo, and create a PR in the bundler repo ?

**Comment by veecue** [comment]
I don't think, that Linux executables support icons. If you want to create .desktop files, I really would go that step and create complete packages with proper prefixes (for example /usr/bin/GroupMatcher for the executable and /usr/share/applications for the .desktop). Then we also could use electron as a package dependency instead of the provisioner thing. What do you think? Windows is a completly different story however... :D

**Comment by asticode** [comment]
Mmm I'd rather keep the same overall logic for every os and keep using the provisioner thingy. Creating a `.desktop` for linux however seems like a good idea. Feel free to create a PR to add this.

**Comment by asticode** [comment]
@veecue did you have time to work on adding the `go-astilectron-bindata` features to the bundler? If not, do you mind if I start working on it this week-end? I won't have the time to finish it, but I want to be able to test everything in action this week-end.

**Comment by asticode** [comment]
@veecue I've added the astilectron-bindata logic to the [bundler](https://github.com/asticode/go-astilectron-bundler). You can check it out and let me know if I've missed anything.

You seem familiar with the Linux `.desktop` extension therefore feel free to submit a PR for this TODO: https://github.com/asticode/go-astilectron-bundler/blob/1d47c8fd03bdedd0892779043019f9b2f8d557ee/bundler.go#L400. 

The idea is to provide a `.desktop` file in addition to the binary.

**Comment by veecue** [comment]
I think that we can close this issue now :) I have created a few new ones on the bundler
